File: src/RuleEngine/Variable.cs
using System;
using System.Collections.Generic;
using System.Linq.Expressions;
using System.Text;

namespace RuleEngine
{
    public class Variable : IRuleElement
    {
        public Variable(string name, object value)
        {
            Name = name;
            Value = value;
        }

        public Guid Id { get; set; } = Guid.NewGuid();
        public string Name { get; set; }
        public object Value { get; set; }

        public bool Process(IRuleElement variable, Operator @operator)
        {
            switch (@operator.Value)
            {
                case Operators.NotEqual:
                    return NotEqual(Value, variable.Value);
                case Operators.Equal:
                    return Equal(Value, variable.Value);
                case Operators.GreaterThan:
                    return GreaterThan(Value, variable.Value);
                case Operators.GreaterThanOrEqual:
                    return GreaterThanOrEqual(Value, variable.Value);
                case Operators.LessThan:
                    return LessThan(Value, variable.Value);
                case Operators.LessThanOrEqual:
                    return LessThanOrEqual(Value, variable.Value);
                case Operators.Contains:
                    return Contains(Value, variable.Value);
                default:
                    throw new ArgumentOutOfRangeException(nameof(@operator));
            }
        }

        private bool NotEqual<T>(T lhs, object rhs)
        {
            return Comparer<T>.Default.Compare(lhs, (T)rhs) != 0;
        }

        private bool Equal<T>(T lhs, object rhs)
        {
            return Comparer<T>.Default.Compare(lhs, (T)rhs) == 0;
        }

        private bool GreaterThan<T>(T lhs, object rhs)
        {
            return Comparer<T>.Default.Compare(lhs, (T)rhs) > 0;
        }

        private bool GreaterThanOrEqual<T>(T lhs, object rhs)
        {
            return Comparer<T>.Default.Compare(lhs, (T)rhs) >= 0;
        }

        private bool LessThan<T>(T lhs, object rhs)
        {
            return Comparer<T>.Default.Compare((T)rhs, lhs) < 0;
        }

        private bool LessThanOrEqual<T>(T lhs, object rhs)
        {
            return Comparer<T>.Default.Compare((T)rhs, lhs) <= 0;
        }

        private bool Contains<T>(T lhs, object rhs)
        {
            return lhs.ToString().Contains(((T)rhs).ToString());
        }
    }
}

File: src/RulesEngine.WebApp/Scripts/rules-engine/Variable.ts
import { IRuleElement } from "./IRuleElement"
import { Operator } from "./Operator"
import { Operators } from "./Operators"

export class Variable implements IRuleElement {
    constructor(name: string, value: any) {
        this.Name = name;
        this.Value = value;
    }

    public Id: string = "";
    public Name: string = "";
    public Value: any = {};

    Process(element: IRuleElement, operator: Operator): boolean {
        switch (operator.Value) {
            case Operators.NotEqual:
                return this.NotEqual(<any>this.Value, <any>element.Value);
            case Operators.Equal:
                return this.Equal(<any>this.Value, <any>element.Value);
            case Operators.GreaterThan:
                return this.GreaterThan(<any>this.Value, <any>element.Value);
            case Operators.GreaterThanOrEqual:
                return this.GreaterThanOrEqual(<any>this.Value, <any>element.Value);
            case Operators.LessThan:
                return this.LessThan(<any>this.Value, <any>element.Value);
            case Operators.LessThanOrEqual:
                return this.LessThanOrEqual(<any>this.Value, <any>element.Value);
            case Operators.Contains:
                return this.Contains(<any>this.Value, <any>element.Value);
            default:
                throw new TypeError("Invalid operator supplied.");
        }
    }

    private NotEqual(lhs: any, rhs: any): boolean {
        return lhs !== rhs;
    }

    private Equal(lhs: any, rhs: any): boolean {
        return lhs === rhs;
    }

    private GreaterThan(lhs: any, rhs: any): boolean {
        return rhs > lhs;
    }

    private GreaterThanOrEqual(lhs: any, rhs: any): boolean {
        return rhs >= lhs;
    }

    private LessThan(lhs: any, rhs: any): boolean {
        return rhs < lhs;
    }

    private LessThanOrEqual(lhs: any, rhs: any): boolean {
        return rhs <= lhs;
    }

    private Contains(lhs: any, rhs: any): boolean {
        return lhs.toString().includes(rhs.toString());
    }
}